servlet-core/src/main/java/io/micronaut/servlet/http/ServletHttpHandler.java

```diff
@@ -342,136 +343,147 @@ private void encodeResponse(ServletExchange<REQ, RES> exchange,
+                    if (response.getContentType().isEmpty()) {
+                        response.contentType(MediaType.APPLICATION_JSON);
+                    }
+                    try (BufferedWriter writer = servletResponse.getWriter()) {
+                        writer.write(body.toString());
+                        writer.flush();
+                    } catch (IOException e) {
+                        throw new HttpStatusException(HttpStatus.INTERNAL_SERVER_ERROR, e.getMessage());
+                    }
+                } else if (body instanceof byte[] byteArray) {
                     try (OutputStream outputStream = servletResponse.getOutputStream()) {
-                        codec.encode(body, outputStream);
+                        outputStream.write(byteArray);
                         outputStream.flush();
-                    } catch (Throwable e) {
-                        throw new CodecException("Failed to encode object [" + body + "] to content type [" + mediaType + "]: " + e.getMessage(), e);
+                    } catch (IOException e) {
+                        throw new HttpStatusException(HttpStatus.INTERNAL_SERVER_ERROR, e.getMessage());
+                    }
+                } else if (body instanceof Writable writable) {
+                    try (OutputStream outputStream = servletResponse.getOutputStream()) {
+                        writable.writeTo(outputStream, response.getCharacterEncoding());
+                        outputStream.flush();
+                    } catch (IOException e) {
+                        throw new HttpStatusException(HttpStatus.INTERNAL_SERVER_ERROR, e.getMessage());
                     }
                 } else {
-                    throw new CodecException("No codec present capable of encoding object [" + body + "] to content type [" + mediaType + "]");
+                    final MediaTypeCodec codec = mediaTypeCodecRegistry.findCodec(mediaType, bodyType).orElse(null);
+                    if (codec != null) {
+                        try (OutputStream outputStream = servletResponse.getOutputStream()) {
+                            codec.encode(body, outputStream);
+                            outputStream.flush();
+                        } catch (Throwable e) {
+                            if (e instanceof CodecException codecException) {
+                                throw codecException;
+                            }
+                            throw new CodecException("Failed to encode object [" + body + "] to content type [" + mediaType + "]: " + e.getMessage(), e);
+                        }
+                    } else {
+                        throw new CodecException("No codec present capable of encoding object [" + body + "] to content type [" + mediaType + "]");
+                    }
                 }
             }
+            responsePublisherCallback.accept(response);
+        } catch (CodecException e) {
+            if (e.getCause() instanceof EOFException) {
```

Comment:
for some reason an EofException occurs only on Jetty during HTTP 1.1 -> 2.0 upgrade protocol handling. This is a workaround for that issue.